File: sort/QuickSort.go
package main

import (
	"fmt"
	"math/rand"
)

func QuickSort(arr []int, start, end int) {
	if start < end {
		key := arr[(start+end)/2]
		i, j := start, end

		for i <= j {
			for arr[i] < key {
				i++
			}
			for arr[j] > key {
				j--
			}

			if i <= j {
				arr[i], arr[j] = arr[j], arr[i]
				i++
				j--
				fmt.Println(arr)
			}
		}

		if start < j {
			QuickSort(arr, start, j)
		}

		if end > i {
			QuickSort(arr, i, end)
		}
	}
}

func qsort(arr []int, start, end int) {
	if start < end {
		key := arr[rand.Intn(len(arr)-1)]
		i, j := start, end

		for arr[i] < key {
			i++
		}
		for arr[j] > key {
			j--
		}
		if i <= j {
			arr[i], arr[j] = arr[j], arr[i]
			i++
			j--
		}
		if start < j {
			qsort(arr, start, j)
		}
		if end > i {
			qsort(arr, i, end)
		}
	}
}

func main() {
	//arr := []int{70,75,69,32,88,18,16,58}
	//QuickSort(arr,0,len(arr) - 1)
	//fmt.Println(arr)

	arr := []int{70, 75, 69, 32, 88, 89, -456, 23, 0, 45, 654, 23, 7, 52, 76, 18, 16, 58}
	array := []int{19, 2, 3, 4, 56, 43, -3, 6}

	qsort(arr, 0, len(arr)-1)
	qsort(array, 0, len(array)-1)
	//	//array := []int{3, 7, 9, 8, 38, 93, 12, 222, 45, 93, 23, 84, 65, 2}
	//
	//	//qsort(array,0,len(array) - 1 )
	fmt.Println(arr)
	//	//fmt.Println(array)
	fmt.Println(array)

}

File: java/structure/QuickSort.java
import java.util.Random;
import java.util.Arrays;

public class QuickSort {
    public static void main(String[] args) {
        int []arr = {34,678,-57,768,0,67,895,24,789,234,6,348,357};
        qsort(arr,0,arr.length - 1);
        System.out.println(Arrays.toString(arr));
    }

    public static void qsort(int []arr,int start,int end){
        if (start < end){
            int i = start;
            int j = end;
            Random rand = new Random();
            int length = arr.length;
            int key = arr[rand.nextInt(length)];

            while(arr[i] < key){
                i++;
            }

            while(arr[j] > key){
                j--;
            }

            if (i <= j){
                swap(arr,i,j);
                i++;
                j--;
            }

            if (start < j){
                qsort(arr,start,j);
            }
            if (end > i){
                qsort(arr,i,end);
            }
        }
    }

    private static void swap(int []arr,int i,int j){
        int temp = arr[i];
        arr[i] = arr[j];
        arr[j] = temp;
    }
}